### PR TITLE
Display "Go to latest stable" for pre-release

### LINKED ISSUE #1044: Show different message than 'Go to latest version' on pre-releases
We automatically redirect to the first non-prerelease version when available, but on pre-release versions we still display the "Go to latest version" box in the tab, see [sqlx v0.4.0-beta.1](https://docs.rs/sqlx/0.4.0-beta.1/sqlx/index.html)

### PR DESCRIPTION
**Pull request by robinhundt**
closes #1044

### CODE REVIEW

File: src/web/rustdoc.rs [hunk 1]
```diff
@@ -367,6 +368,10 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
     // Get the latest version of the crate
     let latest_version = latest_release.version.to_string();
     let is_latest_version = latest_version == version;
+    let is_prerelease = semver::Version::parse(&version)
+        // should be impossible unless there is a bug in match_version
+        .map_err(|_| Nope::InternalServerError)?
+        .is_prerelease();
```
Comment (jyn514) [inline src/web/rustdoc.rs:374]: I don't think this is true, we allow exact matches before checking semver: https://github.com/rust-lang/docs.rs/blob/fb06f95e4a8537fa321c3b192a73ffa6dbd68e97/src/web/mod.rs#L333
```suggestion
    // If this is an exact match (not a valid semver), assume it's not a stable version.
    let is_prerelease = semver::Version::parse(&version).unwrap_or(true);
```

File: templates/rustdoc/header.html [hunk 2]
```diff
@@ -191,6 +191,16 @@
                             </a>
                         </li>
 
+                    {# If the crate is a pre-release, display a warning and offer a link to the latest stable release#}
+                    {%- elif is_prerelease -%}
+                    <li class="pure-menu-item">
+                        <a href="{{ latest_path | safe }}" class="pure-menu-link warn"
+                           title="You are seeing a pre-release version of {{ krate.name }} crate. Click here to go to latest stable version.">
```
Comment (jyn514) [inline templates/rustdoc/header.html:198]: I don't think this is quite right ... You could be on a pre-release version with no stable version available. I think this block should be inside `{%- elif not is_latest_version -%}` and only change the text, not the behavior.

File: src/web/rustdoc.rs [hunk 3]
```diff
@@ -367,6 +368,10 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
     // Get the latest version of the crate
     let latest_version = latest_release.version.to_string();
     let is_latest_version = latest_version == version;
+    let is_prerelease = semver::Version::parse(&version)
+        // should be impossible unless there is a bug in match_version
+        .map_err(|_| IronError::new(Nope::InternalServerError, status::InternalServerError))?
```
Comment (jyn514) [inline src/web/rustdoc.rs:373]: Hmm, I think we could `impl From<Nope> for IronError` to make this easier ... probably should go in a separate PR though.

File: src/web/rustdoc.rs (hunk 1 above)
Comment (robinhundt) [reply src/web/rustdoc.rs]: Oh, you're right. But the suggested change won't work, because then `GET /sqlx/0.3` would be treated as a pre-release although the semver compatible version 0.3.5 would be returned.  
I thought just parsing the Version here would be the least invasive change, but it seems like I'll have to get the actual version returned from the DB for the release serverd.

Comment (robinhundt) [reply src/web/rustdoc.rs]: Mhh, i looked at it again. The `version` is actually guaranteed to be semver compatible, because in the case of an exact match the version is semver compatible (if not, the db contains an invalid semver version) and in the case of an inexact match `rustdoc_html_server_handler` will redirect to the exact version.
https://github.com/rust-lang/docs.rs/blob/c444b639dd489a8985426217feea5b1ec8d97c2c/src/web/rustdoc.rs#L303-L310

Comment (jyn514) [reply src/web/rustdoc.rs]: I don't follow, `"0.3"` is a valid semver and would be parsed successfully.

Comment (jyn514) [reply src/web/rustdoc.rs]: > if not, the db contains an invalid semver version

Hmm ... I don't think we enforce semantic versioning in the database, we depend on cargo to do it for us. In principle a 500 is probably the right thing to do, but it seems a shame to return that when it's so easy to return the page with a different text instead.

File: templates/rustdoc/header.html (hunk 2 above)
Comment (robinhundt) [reply templates/rustdoc/header.html:198]: `{%- elif not is_latest_version -%}` is already the condition for the normal `Go to latest` message.  
I think 
```
{# If the crate is a pre-release, display a warning and offer a link to the latest stable
release. If a crate only has pre-release the warning won't be displayed for the latest #}
{%- elif is_prerelease and not is_latest_version -%}
```
is what we want. This won't display a warning if the user views the latest pre-release for a crate that has no stable releases but displays it for all other pre-releases.

File: src/web/rustdoc.rs (hunk 3 above)
Comment (robinhundt) [reply src/web/rustdoc.rs:373]: Yep, I had the same thought :smile:

File: templates/rustdoc/header.html (hunk 2 above)
Comment (jyn514) [reply templates/rustdoc/header.html]: Yes, that sounds right. If you put it inside `{%- elif not is_latest_version -%}` you only have to change the text displayed without having to copy the HTML around it.

File: src/web/rustdoc.rs (hunk 3 above)
Comment (jyn514) [reply src/web/rustdoc.rs:374]: Opened #1046.

File: templates/rustdoc/header.html (hunk 2 above)
Comment (robinhundt) [reply templates/rustdoc/header.html]: I haven't used Tera a lot. I know how to use the condition for the text inside the span but I don't know how to do that elegantly for the title attribute. I'm also not sure if it would make the template cleaner if I somehow used interpolation for the title.

Comment (jyn514) [reply templates/rustdoc/header.html]: Like this:
```html
                    {# If this isn't the most recent release, offer a link to the latest #}
                    {%- elif not is_latest_version -%}
                      {%- if not is_latest_version -%}
                        {% set title = "You are seeing an outdated version of the " ~ {{ krate.name }} ~ "crate. Click here to go to latest version."
                      {%- else %}
                        {% set title = "You are seeing a pre-release version of the " ~ {{ krate.name }} ~ "crate. Click here to go to latest stable version."
                      {%- endif %}
                        <li class="pure-menu-item">
                            <a href="{{ latest_path | safe }}" class="pure-menu-link warn"
                                title="{% title %}">
                                {{ "exclamation-triangle" | fas(fw=true) }}
                                <span class="title"> Go to latest version</span>
                            </a>
                        </li>
                    {%- endif -%}
```

Comment (jyn514) [reply templates/rustdoc/header.html]: but if you think it's cleaner the other way, that's fine too, I have a tendency to DRY too much sometimes

Comment (robinhundt) [reply templates/rustdoc/header.html]: I've changed it and included the `not is_latest_version and krate.yanked` variant as well.

File: templates/rustdoc/header.html [hunk 4]
```diff
@@ -181,23 +181,24 @@
...
+                    {# If this isn't the most recent stable release, offer a link to the latest #}
                     {%- elif not is_latest_version -%}
+                        {%- if krate.yanked -%}
+                            {%- set tooltip = "You are seeing a yanked version of the " ~ krate.name ~ "crate. Click here to go to latest version." -%}
+                            {%- set title = "This release has been yanked, go to latest version" -%}
+                        {%- elif is_prerelease -%}
+                            {%- set tooltip = "You are seeing a pre-release version of the " ~ krate.name ~ "crate. Click here to go to latest stable version." -%}
+                            {%- set title = "Go to latest stable release" -%}
+                        {%- else -%}
+                            {%- set tooltip = "You are seeing an outdated version of the " ~ krate.name ~ "crate. Click here to go to latest version." -%}
```
Comment (jyn514) [inline templates/rustdoc/header.html:193]: nit: grammar
```suggestion
                            {%- set tooltip = "You are seeing a yanked version of the " ~ krate.name ~ "crate. Click here to go to the latest version." -%}
                            {%- set title = "This release has been yanked, go to latest version" -%}
                        {%- elif is_prerelease -%}
                            {%- set tooltip = "You are seeing a pre-release version of the " ~ krate.name ~ "crate. Click here to go to the latest stable version." -%}
                            {%- set title = "Go to latest stable release" -%}
                        {%- else -%}
                            {%- set tooltip = "You are seeing an outdated version of the " ~ krate.name ~ "crate. Click here to go to the latest version." -%}
```

File: src/web/rustdoc.rs (hunk 1 above)
Comment (robinhundt) [reply src/web/rustdoc.rs]: Would you rather add a new `Nope` variant for this case? I feel like in this case it should be fine to just return a 500. We do the same in the `match_version` func https://github.com/rust-lang/docs.rs/blob/c444b639dd489a8985426217feea5b1ec8d97c2c/src/web/mod.rs#L348-L360

File: src/web/rustdoc.rs [hunk 5]
```diff
@@ -367,6 +368,11 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
     // Get the latest version of the crate
     let latest_version = latest_release.version.to_string();
     let is_latest_version = latest_version == version;
+    let is_prerelease = semver::Version::parse(&version)
+        // should be impossible unless there is a semver incompatible version in the db
+        // Note that there is a redirect earlier for semver matches to the exact version
+        .map_err(|_| IronError::new(Nope::InternalServerError, status::InternalServerError))?
```
Comment (jyn514) [inline src/web/rustdoc.rs:374]: Can you add a log message here that says the error was invalid semver? That would help us notice these bugs.

**Review by jyn514 — APPROVED** [review]
LGTM with nits fixed

File: src/web/rustdoc.rs (hunk 1 above)
Comment (jyn514) [reply src/web/rustdoc.rs]: Hmm, the `Nope` error would just turn into 'InternalServerError', right? I think that would be useful after #1046 to print the log message automatically (in the `from` impl) but other than that I don't think it buys us much.

**Comment by jyn514** [comment]
Thanks for working on this (and putting up with all my nits)!

**Comment by robinhundt** [comment]
No problem :) The code is better that way and you're reviewing really fast :smile: :+1:
